File: Week_10/Practicum/Number of Provinces.py
def dfs(isConnected, visited, current):
    if current not in visited or visited[current]:
        return

    visited[current] = True

    for node in range(len(isConnected[current])):
        if isConnected[current][node]:
            dfs(isConnected, visited, node)


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        visited = {node: False for node in range(len(isConnected))}

        total = 0

        for node in range(len(isConnected)):
            if not visited[node]:
                total += 1
                dfs(isConnected, visited, node)

        return total

File: Week_10/Practicum/Number of Provinces.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& isConnected, int curr, vector<bool>& visited) {
        visited[curr] = true;

        for (int j = 0; j < isConnected.size(); j++) {
            if (isConnected[curr][j] == 1 && !visited[j]) {
                dfs(isConnected, j, visited);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        int counter = 0;

        vector<bool> visited(n, false);

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                counter++;
                dfs(isConnected, i, visited);
            }
        }
        return counter;
    }
};